rfcs/text/0352-graph-type-system.md

```diff
@@ -0,0 +1,2659 @@
+- `deletePropertyType`
+- `getPropertyType`
+- `aggregatePropertyTypes`
+
+- `createEntityType`
+- `updateEntityType`
+- `deleteEntityType`
+- `getEntityType`
+- `aggregateEntityTypes`
+
+- `createLinkType`
+- `updateLinkType`
+- `deleteLinkType`
+- `getLinkType`
+- `aggregateLinkTypes`
+
+- `aggregateDataTypes`
+
+The main changes imposed by the proposed system are that Entity Types must be defined as previously outlined - with canonical Property Type URIs and that new messages for managing Property Types and Link Types must be added.
+
+In the proposed system, Entity Types can be created and updated with the same semantics but have to conform to the [Entity Type](#entity-types-1) schema instead of arbitrary JSON schemas.
+Property Types will be defined similarly to Entity Types in the proposed system, conforming to the [Property Type](#property-types-1) schemas instead.
+Link Types do not contain further structural data, which make them semantically simpler than Entity Types and Property Types.
+
+**Entity Type Create and Update messages in the current system:**
+
+```json
+{
+  // createEntityType message
+  "schema": {
+    "type": "object",
+    "name": "Person",
+    "properties": {
+      "name": { "type": "string" },
+      "email": { "type": "string" },
+      "phoneNumber": { "type": "string" }
+    }
+  }
+}
+```
+
+```json
+{
+  // updateEntityType message
+  "entityTypeId": "Person",
+  "schema": {
+    "properties": {
+      "birthDate": { "type": "string", "format": "date" }
+    }
+  }
+}
+```
+
+**Entity Type Create and Update messages in the proposed system:**
+
+```json
+{
+  // createEntityType message
+  "schema": {
+    "type": "object",
+    "kind": "entityType",
+    "name": "Person",
+    "properties": {
+      "https://blockprotocol.org/types/@alice/property-type/name": {
+        "$ref": "https://blockprotocol.org/types/@alice/property-type/name"
+      },
+      "https://blockprotocol.org/types/@alice/property-type/email": {
+        "$ref": "https://blockprotocol.org/types/@alice/property-type/email"
+      },
+      "https://blockprotocol.org/types/@alice/property-type/phone-number": {
+        "$ref": "https://blockprotocol.org/types/@alice/property-type/phone-number"
+      }
+    }
+  }
+}
+```
+
+```json
+{
+  // updateEntityType message
+  "entityTypeId": "Person",
+  "schema": {
+    // The properties here are partially applied to the original Entity Type.
+    "properties": {
+      "https://blockprotocol.org/types/@alice/property-type/birth-date": {
+        "$ref": "https://blockprotocol.org/types/@alice/property-type/birth-date"
+      }
+    }
+  }
+}
+```
+
+The update messages of both the current and new systems make use of partial schemas, merging the schema given in the message contents with the existing Entity Type. This may or may not be the desired semantics of updating, and could lead to undesired behavior. In that case, the semantics can be changed to treat updates as a complete replacement.
```

Comment:
By whom, how? Do we mean on a per-operation basis, or that we might change the proposal?

Reply:
I think the original text is a little misleading. The semantics around updating is sort of parallel to inheritance/forking, so while this proposal might not change much, the RFC that will address inheritance/forking will bring in the considerations around updating types.